### PR TITLE
Refactor and New Features.

### PR DESCRIPTION
**Pull request by takwas**
Kindly review the changes so far. I am not done with the tests yet.

----
#### Changelog:
- **More modularized**
- **Configuration** - Load configuration from a file. Configuration includes modes: _dev_, _test_ and _deploy_, depending on the running mode of the application. Some values can also be loaded from environment variables.
- **App Logging** - Added logging to the running application
- **Class Logging now done using the Python `logging` module.**
- **Data dir** - A new "data/" directory where application data is saved. "links.json" is also moved to this folder.
- **Dot-prefixed commands.** `.next`, `.pingall`, `.clearqueue` ...
- **Question queue commands**:
 * `!` for _join queue_;
 * `!!` or `!-` for _leave queue_ **(solves issue #17 )**
- **Certain features/commands are optional;** (can be set through configuration.)
- **User object** - Now includes an `IRCUser` class parsed from a user's IRC hostmask.
- **QuestionQueue object** - a custom queue implementation for queueing question indicators.
- **Set channel topic** - Bot can set topic of the channel during a class session. Desired topic is given as argument with the `startclass` command. E.g `startclass This is today's topic.`
- **Unit testing** - Some unit tests included. **(partially solves issue #12)**
- **Fpaste** uploads a log once, and saves the URL for subsequent requests.

- And more...

### CODE REVIEW

**Comment by sayanchowdhury** [comment]
Looks good to me so far. I will pull the changes and test on my machine

**Comment by takwas** [comment]
Good to hear. I'll keep sending PRs as I discover bugs. However, I still need some guide with writing tests. So any tips on that would be appreciated.

**Comment by takwas** [comment]
@sayanchowdhury Some tests have been added. Kindly review.

File: ekan0ra/bot.py [hunk 1]
```diff
@@ -0,0 +1,448 @@
...
+
+        elif msg in ('!', '!-', '!!') and not self.islogging:
+            self.say(
+                self.channel,
+                '%s: No session is going on, feel free to ask a question. '
+                    'You do not have to type %s' % (user.nick, msg)
+            )
+        # end processing question indicator   
+
+        elif msg == '.givemelogs':
```
Comment (sayanchowdhury) [inline ekan0ra/bot.py]: Can you add this to config too? If `False` then this feature is disabled

File: ekan0ra/bot.py [hunk 2]
```diff
@@ -0,0 +1,448 @@
...
+            if not self.last_log_filename:
+                self.say(user.nick, 'Sorry, I do not have the last log.')
+                application_logger.warning('Could not find last class log!')
+            else:
+                sys.argv = ['fpaste', self.last_log_filename]
```
Comment (sayanchowdhury) [inline ekan0ra/bot.py]: One thing that I have in mind for sometime is that if the class is over then maybe we can store the fpaste url. Also I had plans that a nick can only request twice during a session.

File: ekan0ra/bot.py [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+            for command, help_txt in commands.iteritems():
+                self.say(
+                    user.nick,
+                    help_template.format(
+                        command=command, help_text=help_txt
+                    )
+                )      
+
+        elif msg.startswith('.link'):
+            self.links_for_key(msg)
```
Comment (sayanchowdhury) [inline ekan0ra/bot.py]: one thing that can be added here is that the option to list all the links (for admins only) in pm

File: config.py [hunk 4]
```diff
@@ -1,4 +1,135 @@
...
+    #LOG_FILENAME = '' # Override log filename
+    #CLASS_LOGGER_FORMAT = '' # Override format_str for class logger
+    #LOG_PREFIX = '' # Override prefix format_str for class logger
+    #LINKS_FILE = '' # Override links file
+    #SESSION_START_MSG = '' # Override session start message
+    #SESSION_END_MSG = '' # Override session end message
+    #BASE_TOPIC = '' # Override channel default topic
+
+
+    BOTNICK = os.getenv('LOGBOT_NICK', 'sawkat')
```
Comment (sayanchowdhury) [inline config.py]: Why not have a config file rather than setting env variable?

Comment (takwas) [inline config.py]: @sayanchowdhury I thought if someone were to fork the project, they'd find it more flexible to configure.

File: ekan0ra/bot.py (hunk 3 above)
Comment (takwas) [inline ekan0ra/bot.py]: Do you mean, have the `.link` command to be usable only by an admin or its argument `-l` only available to an admin?

**Pull request merged**